garden_ai/mlmodel.py

```diff
@@ -16,90 +17,150 @@ class ModelUploadException(Exception):
+    user_email (str): The email address of the user uploading the model.
+
+    """
+
+    model_name: str = None
+    version: str = None
+    user_email: str = None
+    flavor: str = None
+    connections: list[DatasetConnection] = Field(default_factory=list)
+    namespaced_model_name: str = None
+    model_uri: str = None
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.namespaced_model_name = f"{self.user_email}-{self.model_name}"
+        self.model_uri = f"{self.namespaced_model_name}/{self.version}"
+
+
+def upload_to_model_registry(local_model: LocalModel) -> RegisteredModel:
     """Upload a model to Garden-AI's MLflow model registry.
 
     Parameters
     ----------
-    model_path : str
-        The file path specific to the model type to be uploaded, e.g. ``/path/to/my/model.pkl``
-        or ``/path/to/my/tf_model/`` or ``/path/to/my/torch_model.pt``.
-    model_name : str
-        The name under which to register the uploaded model.
-    user_email : str
-        The email address of the user uploading the model.
-    flavor : str
-        The library the model was made with, e.g. ``sklearn`` or ``tensorflow``.
-    extra_pip_requirements : List[str], optional
-        A list of additional pip requirements needed to load and/or run the model.
-        Defaults to None.
+    local_model : LocalModel
+        The model to upload.
 
     Returns
     -------
-    str
-        The full name and version number of the uploaded model in the MLflow
-        registry, which can be used to fetch the model with a call to ``Model(...)``.
+    RegisteredModel
+        Includes the full model_uri, which can be used to fetch the model with a call to ``Model(...)``.
 
     Raises
     ------
     ModelUploadException
         If an error occurs during the upload process, such as failure to open or
         parse the model, or failure to retrieve the latest version of the model.
     """
-    full_model_name = f"{user_email}-{model_name}"
+    _push_model_to_registry(local_model)
+    return _assemble_metadata(local_model)
+
+
+def _push_model_to_registry(local_model: LocalModel):
+    flavor, local_path = local_model.flavor, local_model.local_path
     try:
-        if flavor == "sklearn" and pathlib.Path(model_path).is_file:
-            with open(model_path, "rb") as f:
+        if flavor == "sklearn" and pathlib.Path(local_path).is_file:
+            with open(local_path, "rb") as f:
                 loaded_model = pickle.load(f)
-                mlflow.sklearn.log_model(
-                    loaded_model,
-                    user_email,
-                    registered_model_name=full_model_name,
-                    extra_pip_requirements=extra_pip_requirements,
-                )
-        elif flavor == "tensorflow" and pathlib.Path(model_path).is_dir:
+                log_model_variant = mlflow.sklearn.log_model
+        elif flavor == "tensorflow" and pathlib.Path(local_path).is_dir:
```

Comment:
looks like validators fire in order of appearance of their fields, and can include a `values` argument to get access to previously-validated fields -- [example](https://docs.pydantic.dev/latest/usage/validators/) from the docs

Reply:
Oh no I agree, I just meant that a validator could enforce that the model_path string looks like a `directory/with/model` when flavor is pytorch vs looks like a `path/to/model.pkl` when flavor is sklearn, etc. That way, we'd only ever need `if flavor == "tensorflow": ...` instead of remembering to check both `if flavor == "tensorflow" and pathlib.Path(local_path).is_dir: ...`

Reply:
I think this is where I'm confused. As I understand it, `pathlib.Path(local_path).is_dir` is checking the filesystem. So you could do some kind of check in the validator for file endings maybe, but you would still need to check later that the file/directory is indeed a file/directory on the filesystem. I just don't see a lot of value in an upfront check that could only use heuristics.